## MNIST classification with Dense Neural Network

### Load Required Libraries

In [1]:
from keras.callbacks import ModelCheckpoint
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.utils import np_utils
import numpy as np

Using TensorFlow backend.


### Load data

In [2]:
BATCH_SIZE =128
epochs = 10

(xtrain, ytrain), (xtest, ytest) = mnist.load_data()

## reshape each sample to a vector containing 784 points

xtrain = xtrain.reshape(60000, 784).astype("float32")/255
xtest = xtest.reshape(10000, 784).astype("float32")/255

## convert label to one hot encoding

ytrain = np_utils.to_categorical(ytrain, 10)
ytest =np_utils.to_categorical(ytest, 10)


### Create Sequential model with Dense and Dropout layers

In [3]:
model = Sequential()

model.add(Dense(512, input_shape=(784,), activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(512, activation ="relu"))
model.add(Dropout(0.2))
model.add(Dense(10, activation="softmax"))

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 512)               401920    
_________________________________________________________________
dropout_1 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 512)               262656    
_________________________________________________________________
dropout_2 (Dropout)          (None, 512)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                5130      
Total params: 669,706
Trainable params: 669,706
Non-trainable params: 0
_________________________________________________________________


#### Compile and train model

In [4]:
model.compile(optimizer="rmsprop", loss="categorical_crossentropy", metrics=['accuracy'])

In [5]:
history= model.fit(xtrain, ytrain, epochs= 10, verbose=1, validation_split=0.25)

Train on 45000 samples, validate on 15000 samples
Epoch 1/10
45000/45000 [==============================] - 28s 625us/step - loss: 0.2418 - accuracy: 0.9281 - val_loss: 0.1709 - val_accuracy: 0.9569
Epoch 2/10
45000/45000 [==============================] - 28s 617us/step - loss: 0.1294 - accuracy: 0.9649 - val_loss: 0.1332 - val_accuracy: 0.9683
Epoch 3/10
45000/45000 [==============================] - 28s 620us/step - loss: 0.1142 - accuracy: 0.9723 - val_loss: 0.1355 - val_accuracy: 0.9696
Epoch 4/10
45000/45000 [==============================] - 28s 620us/step - loss: 0.1036 - accuracy: 0.9756 - val_loss: 0.1524 - val_accuracy: 0.9693
Epoch 5/10
45000/45000 [==============================] - 27s 605us/step - loss: 0.1007 - accuracy: 0.9778 - val_loss: 0.1495 - val_accuracy: 0.9758
Epoch 6/10
45000/45000 [==============================] - 28s 621us/step - loss: 0.0982 - accuracy: 0.9794 - val_loss: 0.1986 - val_accuracy: 0.9716
Epoch 7/10
45000/45000 [==============================] 

In [6]:
score = model.evaluate(xtest, ytest, verbose=1)
print(score)

10000/10000 [==============================] - 2s 186us/step
[0.19687760202035776, 0.977400004863739]


With just dense and dropout layers we can achieve a very high accuracy of 97.7%. But with the help of convolution layers and image augmentation technique accuracy can be increased.